Stripe checkout customization cookbook (#9880)
Sylius/Sylius: docs/cookbook/payments/stripe-customization.rst

How to customize Stripe Credit Card payment?
============================================

Prerequisite: install Stripe
----------------------------

First of all, Stripe needs to be installed, configured and tested, using this :doc:`documentation</cookbook/payments/stripe>`.

Override Payum templates
------------------------

.. note::

    By doing this you'll be able to use a custom layout and to pass extra parameters to Stripe like the user email.

.. code-block:: yaml

    # app/config/config.yml
    twig:
        paths:
            "%kernel.root_dir%/../app/Resources/PayumCore/views": PayumCore
            "%kernel.root_dir%/../app/Resources/PayumStripe/views": PayumStripe

> Having this configuration defined, you can now create your customuzations in your `app/Ressource/` directory, as usual.

Reviewer: ```suggestion
Having this configuration defined, you can now create your customizations in your `app/Resources/` directory, as usual.
```

        {{ parent() }}

        {% block payum_vendor_javascripts "" %}
        {% block payum_javascripts "" %}
    {% endblock %}

For example, you could also pass extra parameters to Stripe:

.. code-block:: twig

    # app/Resources/PayumStripe/views/Action/obtain_checkout_token.html.twig
    {% extends layout ?: "@PayumCore/layout.html.twig" %}

    {% block payum_body %}
        {{ parent() }}

        <form action="{{ actionUrl|default('') }}" method="POST">
            <script
                    {# see https://stripe.com/docs/checkout#integration-simple #}
                    src="https://checkout.stripe.com/checkout.js" class="stripe-button"
                    data-key="{{ publishable_key }}"
                    data-image="{{ asset('path/to/image.png') }}"
                    data-name="{{ model.name|default("Your store name") }}"
                    data-description="{{ model.description|default("") }}"
                    data-amount="{{ model.amount }}"
                    data-currency="{{ model.currency|default("EUR") }}"
                    data-email="{{ app.user.email }}"
                    data-locale="auto"
                    data-allow-remember-me="false"

> 

Reviewer: ```suggestion
```

        </script>
    {% endblock payum_javascripts %}

Add metadata to Stripe
----------------------

The metadata have to be given later to Stripe, so you can't simply use the template overriding, instead, you have to define a Payum Extension that will handle this for you at the right time:

First you have to create a Service

.. code-block:: php

    <?php
    /* src/AppBundle/Payment/StripeAddMetadataOnCaptureExtensions.php */

    namespace AppBundle\Payment;

    use AppBundle\Entity\Customer;
    use AppBundle\Entity\Order;
    use AppBundle\Entity\OrderItem;
    use Payum\Core\Extension\Context;
    use Payum\Core\Extension\ExtensionInterface;
    use Payum\Core\Request\Capture;

    use Sylius\Bundle\PayumBundle\Request\GetStatus;
    use Sylius\Component\Core\Model\PaymentInterface as SyliusPaymentInterface;

    /**
     *
     * This class handles the addition of metadata to Stripe
     *
     * It is wired using Payum's built in extension system.
     * It is wired to Stripe_checkout only using a factory filter on the service tag.
     * It is additionally filtered to the Capture request using a code check.
     *

>      * @author Bruno DA SILVA (https://github.com/bruno-ds)

Reviewer: We usually don't mention authors in codeblocks in our docs.

     */
    class StripeAddMetadataOnCaptureExtensions implements ExtensionInterface
    {
        /**

>          * @var Context $context

Reviewer: Could you remove these kinds of docblocks (as there are typehints already)?

Reviewer 2: Typehints in function body are 👍 IMO, especially in the documentation where you cannot jump to the code and check the return value. This one is redundant due to typehints in function parameters. I would only change them to interfaces, whenever it's possible (e.g. `Order` -> `OrderInterface`).

         */
        public function onPreExecute(Context $context)
        {
            if (!$this->supports($context)) {
                return;
            }

            return;
        }

        /**
         * @var Context $context
         */
        public function onExecute(Context $context)
        {
            if (!$this->supports($context)) {
                return;
            }

>             /** @var $request Capture */

Reviewer: ```suggestion
            /** @var Capture $request */
```
